allensdk/brain_observatory/gaze_mapping/_filter_utils.py

```diff
@@ -0,0 +1,132 @@
+import logging
+import numpy as np
+
+
+def medfilt_custom(x, kernel_size=3):
+    '''This median filter returns 'nan' whenever any value in the kernal width
+    is 'nan' and the median otherwise'''
+    T = x.shape[0]
+    delta = kernel_size // 2
+
+    x_med = np.zeros(x.shape)
+    window = x[0:delta + 1]
+    if np.any(np.isnan(window)):
+        x_med[0] = np.nan
+    else:
+        x_med[0] = np.median(window)
+
+    # print window
+    for t in range(1, T):
+        window = x[t - delta:t + delta + 1]
+        # print window
+        if np.any(np.isnan(window)):
+            x_med[t] = np.nan
+        else:
+            x_med[t] = np.median(window)
+
+    return x_med
+
+
+def median_absolute_deviation(a, consistency_constant=1.4826):
+    '''Calculate the median absolute deviation of a univariate dataset.
+
+    Parameters
+    ----------
+    a : numpy.ndarray
+        Sample data.
+    consistency_constant : float
+        Constant to make the MAD a consistent estimator of the population
+        standard deviation (1.4826 for a normal distribution).
+
+    Returns
+    -------
+    float
+        Median absolute deviation of the data.
+    '''
+    return consistency_constant * np.nanmedian(np.abs(a - np.nanmedian(a)))
+
+
+def post_process_cr(cr_params):
+    """This will replace questionable values of the CR x and y position with
+    'nan'.
+
+        1)  threshold ellipse area by 99th percentile area distribution
+        2)  median filter using custom median filter
+        3)  remove deviations from discontinuous jumps
+
+        The 'nan' values likely represent obscured CRs, secondary reflections,
+        merges with the secondary reflection, or visual distortions due to
+        the whisker or deformations of the eye
+
+    Parameters
+    ----------
+    cr_params: numpy.ndarray
+        (Nx5) array of pupil parameters [x, y, angle, axis1, axis2].
+    """
+
+    area = np.pi * (cr_params.T[3] / 2) * (cr_params.T[4] / 2)
+
+    # compute a threshold on the area of the cr ellipse
+    dev = median_absolute_deviation(area)
+    if dev == 0:
+        logging.warning("Median absolute deviation is 0,"
+                        "falling back to standard deviation.")
+        dev = np.nanstd(area)
+    threshold = np.nanmedian(area) + 3 * dev
+
```

Comment:
feel free to ignore b/c time but much of this function seems like it could be:
1. a separate function
2. applied to the x and y positions without duplicating every line of code

Reply:
Yes, completely agree, unfortunately, this was a copy/paste of the existing implementation as I discovered this filtering functionality late (was spending most of my time understanding the gaze mapping stuff).